src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -50494,6 +50499,226 @@ static void G(object? x)
+        Task.Run(() => Baz());
+    }
+
+    private static object GetNotNullObject() => new object();
+
+    private static bool VerifyNotNull(object instance) => instance != null;
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(44411, "https://github.com/dotnet/roslyn/issues/44411")]
+        public void LambdaWritesDoNotUpdateContainingMethodState()
+        {
+            var source = @"
+using System;
+
+class Program
+{
+    static void M1()
+    {
+        string? x = ""hello"";
+        M2(() =>
+        {
+            x = null;
+        });
+        x.ToString();
+
+        x = null;
+        x.ToString(); // 1
+    }
+
+    static void M2(Action a) { }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (16,9): warning CS8602: Dereference of a possibly null reference.
+                //         x.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(16, 9));
+        }
+
+        [Fact, WorkItem(44411, "https://github.com/dotnet/roslyn/issues/44411")]
+        public void LambdaWritesDoNotUpdateContainingMethodState_02()
+        {
+            var source = @"
+using System;
+
+class Program
+{
+    static void M1()
+    {
+        M2(() =>
+        {
+            string? y = ""world"";
+            M2(() =>
+            {
+                y = null;
+            });
+            y.ToString();
+
+            y = null;
+            y.ToString(); // 1
+        });
+    }
+
+    static void M2(Action a) { }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (18,13): warning CS8602: Dereference of a possibly null reference.
+                //             y.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(18, 13));
+        }
+
+        [Fact, WorkItem(44411, "https://github.com/dotnet/roslyn/issues/44411")]
+        public void AnonymousMethodWritesDoNotUpdateContainingMethodState()
+        {
+            var source = @"
+using System;
+
+class Program
+{
+    static void M1()
+    {
+        string? x = ""hello"";
+        M2(delegate()
+        {
+            x = null;
+        });
+        x.ToString();
+
+        x = null;
+        x.ToString(); // 1
+    }
+
+    static void M2(Action a) { }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (16,9): warning CS8602: Dereference of a possibly null reference.
+                //         x.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(16, 9));
+        }
+
+        [Fact, WorkItem(44411, "https://github.com/dotnet/roslyn/issues/44411")]
+        public void LocalFunctionWritesDoNotUpdateContainingMethodState()
```

Comment:
The behavior of this test gave me pause, particularly due to this comment in the original code:

```cs
// State changes to captured variables are recorded, as calls to local functions
// transition the state of captured variables if the variables have state changes
// across all branches leaving the local function
```

However, the behavior of this test does not differ in master, therefore I am inclined to file a separate bug for it.

Reply:
Which part of the behavior is giving u pause here?

Reply:
The comment claims that calls to local functions will transition the state of captured variables if the variables have state changes across all branches leaving the local function. I interpreted this to mean that e.g. if you unconditionally assign 'null' to a captured variable as in this test, it's expected that the state of the captured variable will change to maybe-null at call sites to the local function.

Reply:
Agree. We should file a follow up issue.

Reply:
Filed #45770